### ISSUE TITLE
Hosting "normal" solutions

### ISSUE THREAD

**Issue by atesgoral** [atesgoral/hrm-solutions#69]
This repo started off with standard solutions I came up with for all levels, without necessarily meeting or exceeding the size and speed challenges. I set it as a goal to make this into a collaborative space to push the boundaries of size/speed optimizations, and it has been going really well so far :) But, I'm also a little bothered by the fact that there's no longer any room for any "normal" solutions that could help people with learning about algorithms and get unstuck in the game.

I can think of 2 options:

Option A:
Continue with hosting just size/speed optimizations.

Option B:
Find a naming convention / folder structure to also host several "normal" solutions (like both a bubble sort and a selection sort alternative to 41 Sorting Floor)

Any suggestions?

**Comment by atesgoral** [comment]
Attn: @unframework @spenserhale @grimer30 @Ferwex @hermanprawiro @IAmWave @valkum @codejnki @Patrick-Jakubowski @lzelus @CozyRocket @peterfreese @polarathene @cowboy @ekx @pyaehtetaung @Otto42 @IllegallyBlind @WolfWings @nanashi-juanto @FireGoblin @AlanDeSmet

**Comment by Otto42** [comment]
A solution set for "normal" would be fine for holding these, I feel. Any "exact" solutions would be ideal for this set, although they could also be made to be the most straightforward approach.

Also, the set of files is getting unwieldy in such a case. Maybe organize them into folders?

Additional: Now that the game has achievements, some solutions that get those achievements could be made as well. There's at least 4 achievements which can be gotten from using specific programs on specific levels.

**Comment by codejnki** [comment]
Could start by creating a folder per level which could act as a container various solutions to the problem.  File names such as _tutorial_ and _step#_ could be used to help show the evolution to a solution.

**Comment by codejnki** [comment]
Maybe a _hints_ file that wouldn't necessarily contain code but instead pointers towards the solution that someone could read first.  For example last night I made it up to floor 39 and my first thought was "hm, where do I even start with this".

**Comment by AlanDeSmet** [comment]
Maintaining a "straightforward" set of solutions seems like a really neat idea.  Multiple solutions with different goals seems like a good idea: Most straightforward/easiest to understand, as straightforward as possible but makes speed par, as straightforward as possible but makes size par.

codejnkl's suggestion for evolution would be more work, but seems really neat. Although perhaps that road should lead to more prose explaining the evolution than relying on people seeing the evolution (doubly so since HRM doesn't provide internal diffing tools).

**Comment by unframework** [comment]
Folders make sense. And there's definitely room for several different solutions within one category (size/speed), so that nobody feels that their work is thrown out altogether, and also to illustrate different approaches.

**Comment by cowboy** [comment]
What about a folder structure like this? It would allow for any number of size and speed variants per puzzle.

* `NN-Name-Of-Puzzle`
 * `size-NN-githubusername.asm`
 * `...`
 * `speed-NN-githubusername.asm`
 * `...`

Eg.

* `22-Fibonacci-Visitor`
 * `size-18-cowboy.asm`
 * `speed-58-Patrick-Jakubowski.asm`
 * `...`

**Comment by Otto42** [comment]
If you want something to get started with, here's an annotated Bubble Sort for 41 Sorting Floor:

**Comment by atesgoral** [comment]
@Otto42 The code snipped didn't come through.

**Comment by atesgoral** [comment]
@cowboy +1 to that folder structure.

For algorithm variants, we could perhaps name the files as:

* 41-Sorting-Floor
 * bubble-githubusername.asm
 * quick-githubusername.asm
 * bozo-githubusername.asm

**Comment by nanashi-juanto** [comment]
I also agree to using a per-level folder structure.

I imagine having up to one each best in size and in speed, and up to two boths - preferably just one that does it better unless the second is different enough to provide a different perspective at solving the problem while meeting or coming close to the first's score*.

*This being said, some time ago a solution of mine replaced 12-Tetracontiplier.size.speed, achieving the same score with less memory. We may retrieve the old solution written by @atesgoral and attach brief explanations (for each problem). I looked at his solution again; the route that he took in order to multiply by 40 was
> set to a, a+a (2n), set to b, b+b+a (5n), set to c, c+c (10n), set to c, c+c+c (20n, 40n)

while I extend the previous solution, which asks to multiply by 8:

> set to a, a+a (2n), set to a, a+a (4n), set to a, a+a (8n), set to a, a+a+a+a (16n-40n, multiples of 8)

**EDIT:** Heck, revise the former. The following is now clean and should be kept:

> set to a, a+a (2n), set to b, b+b+a (5n), set to a, a+a (10n), set to a, a+a+a (20n, 40n)

**Comment by atesgoral** [comment]
Yeah, I think I might have to comb through the commits to cherry pick some nice solutions to save them as separate files...

**Comment by CozyRocket** [comment]
I noticed my contribution was replaced which I am actually glad. Im all
about optimizing the current files, even if they use one less command or
makes it slightly faster is what we should aim for. Also if all the files
had labels that explain what's happening, and possibly replace some of the
text in files that I have seen as couldn't read the text that well.

On other things. I hope they add custom levels.

**Comment by IAmWave** [comment]
Are you interested in solutions that are made just to test the limitations of the language? I was thinking about implementing merge sort.

**Comment by atesgoral** [comment]
@IAmWave Absolutely!

**Comment by atesgoral** [comment]
OK going ahead with splitting stuff into subfolders at the first chance.

**Comment by WolfWings** [comment]
Total agreement on the folder structure, and also agreement: There's "speed even if it doesn't pass the general purpose case" and then there's a general-purpose solver, for some things there's even levels in between and alternative-but-interesting examples like various types of sorting implemented.

**Comment by Otto42** [comment]
@atesgoral The code I posted is in the gist. Copy and paste as you like. Not a pull request. Just a starting point.

**Comment by atesgoral** [comment]
@Otto42 Cool! Saw it now!

**Comment by peterfreese** [comment]
I also wonder its worth having a special name for exploit solutions. For
example, the current best speed for 09-Zero Preservation Initiative is not
a general purpose solution and exposes/relies on there being a limited set
of test inputs.

**Comment by Patrick-Jakubowski** [comment]
Aye, I definitely agree with the folder structure, and I also feel there should be a specific naming convention for solutions that are fastest, but only within the specific case given in the level, not generally in the real world.  For instance, in my fibonacci solution, it only works so quickly because it knows the highest value it needs to compute is 21.  If someone tried using the assembly code to compute all the fibonacci values up to 34, it would fail.

Maybe we should have a different category altogether, as both speed and lines can be different for problems that are general solutions versus ones that get this particular set done.  Definitely easier to do this with subfolders though.

**Comment by atesgoral** [comment]
@Patrick-Jakubowski: @nanashi-juanto has been including the keyword "exploit" in filenames for solutions that use some sort of exploit to beat size/speed records. I don't have a strong preference between using subfolders versus keywords.

**Comment by atesgoral** [comment]
I started thinking it would be nice to include the par numbers in the folder names, such as:

`17-Exclusive-Lounge-12-28/`

It can then be seen at a glance what a solution's standing is relative to the par:

`17-Exclusive-Lounge-12-28/12-27-AaronKnowles.asm`

I'm looking forward to a sweeping file rename to automatically include the size/speed scores in filenames using [hrm-cpu](https://github.com/nrkn/hrm-cpu).

**Comment by nanashi-juanto** [comment]
@atesgoral: This would not only solve the otherwise-duplicate filenaming issue for 41-Sorting-Floor where I attached unofficial tags to separate them; but also quickly provide par information without any need to state which one a solution excels in. I was getting worried and began questioning my renames.

@Patrick-Jakubowski mentioned general input (quality/quantity) and level-specific input set solutions. I spent quite some time thinking about this. I consider using the keyword "specific" and leaving "exploit" for the one currently known super-hacky solution and the likes that ignores both quality and quantity of input:

```
37-Scavenger-Chain.8.63/32.20.exploit-WolfWings.asm
    Copies a pre-determined message to the outbox. This is either ESCAPEAPE or APEESCAPE
    depending on whether 0 or 23 comes up first in the inbox. The second number doesn't matter.
```

In my opinion, I prefer using dots, seen above as an example, for its visual separation between level and par numbers; tags-username.
